### PR TITLE
Add the take_until range parser

### PR DESCRIPTION
**Pull request by brennie**
The `take_until(r)` parser parses until the range `r` is found, at which
point all input up to (but not including) `r` will be returned. If `r`
is not found, an error will be returned.

### CODE REVIEW

File: src/range.rs [hunk 1]
```diff
@@ -182,6 +182,69 @@ pub fn take_while1<I, F>(f: F) -> TakeWhile1<I, F>
...
+                    if xs == self.0 {
+                        break;
+                    } else {
+                        to_consume += 1;
+                        if look_ahead_input.uncons().is_err() {
+                            unreachable!();
+                        }
+                    }
+                },
+                Err(_) => return EmptyErr(ParseError::empty(look_ahead_input.position())),
```
Comment (Marwes) [inline src/range.rs]: Don't ignore the error here (use `ParseError::new`)

File: src/range.rs [hunk 2]
```diff
@@ -182,6 +182,69 @@ pub fn take_while1<I, F>(f: F) -> TakeWhile1<I, F>
...
+            };
+        }
+
+        match input.uncons_range(to_consume) {
+            Ok(x) => ConsumedOk((x, input)),
```
Comment (Marwes) [inline src/range.rs]: If the tag was found immedietly we haven't consumed any input so we need to return `EmptyOk` (to_consume == 0)

File: src/range.rs [hunk 3]
```diff
@@ -209,4 +272,13 @@ mod tests {
         let result = parser.parse("hell\u{00EE} world");
         assert!(result.is_err());
     }
+
+    #[test]
+    fn take_until_test() {
```
Comment (Marwes) [inline src/range.rs]: No need to duplicate the doc test here.

File: src/range.rs [hunk 4]
```diff
@@ -182,6 +182,69 @@ pub fn take_while1<I, F>(f: F) -> TakeWhile1<I, F>
...
+/// # fn main() {
+/// let mut parser = take_until("\r\n");
+/// let result = parser.parse("To: user@example.com\r\n");
+/// assert_eq!(result, Ok(("To: user@example.com", "\r\n")));
+/// let result = parser.parse("Hello, world\n");
+/// assert!(result.is_err());
+/// # }
+/// ```
+#[inline(always)]
+pub fn take_until<I>(r: I::Range) -> TakeUntil<I> where I: RangeStream
```
Comment (Marwes) [inline src/range.rs]: I'd probably call it `take_until_range` to leave the possibility open for `take_while_range` and `take_until` (which would be equivalent to `take_while(|t| !predicate(t))`)

File: src/range.rs [hunk 5]
```diff
@@ -182,6 +182,69 @@ pub fn take_while1<I, F>(f: F) -> TakeWhile1<I, F>
...
+
+        let len = self.0.len();
+        let mut to_consume = 0;
+        let mut look_ahead_input = input.clone();
+    
+        loop {
+            match look_ahead_input.clone().uncons_range(len) {
+                Ok(xs) => {
+                    if xs == self.0 {
+                        break;
```
Comment (Marwes) [inline src/range.rs]: Instead of breaking you could return the result immediately here.

File: src/range.rs (hunk 1 above)
Comment (brennie) [inline src/range.rs]: Isn't the point of `parse_lazy` to add empty errors which can be filled in later? Or am I misunderstanding?

**Comment by brennie** [comment]
I've updated the PR based on your review feedback.

**Comment by Marwes** [comment]
I'm currently abroad for work but I will take a proper look after i get back home (~1 day)!

File: src/range.rs [hunk 6]
```diff
@@ -182,6 +182,72 @@ pub fn take_while1<I, F>(f: F) -> TakeWhile1<I, F>
...
+        let len = self.0.len();
+        let mut to_consume = 0;
+        let mut look_ahead_input = input.clone();
+
+        loop {
+            match look_ahead_input.clone().uncons_range(len) {
+                Ok(xs) => {
+                    if xs == self.0 {
+                        if to_consume == 0 {
+                            return EmptyOk((self.0.clone(), input));
```
Comment (Marwes) [inline src/range.rs]: The parser shouldn't return the tag here but instead an empty range (which we can get from input.uncons_range(0))

Comment (brennie) [inline src/range.rs]: Done & done.

**Comment by Marwes** [comment]
Thanks!

**Pull request merged**